A `frameset` tag inside `template` should be simply ignored

Ahmad-S792 opened the pull request:
#### dc45153ce2d1ba0b2f778a5da7c835d99556feb6
<pre>
A `frameset` tag inside `template` should be simply ignored

<a href="https://bugs.webkit.org/show_bug.cgi?id=256455">https://bugs.webkit.org/show_bug.cgi?id=256455</a>

Reviewed by Ryosuke Niwa.

This patch aligns WebKit with Web-Spec [1] and Blink / Chromium.

Merge: <a href="https://chromium.googlesource.com/chromium/src/+/93198017327e240584facc073ae5853a1cc78a6c">https://chromium.googlesource.com/chromium/src/+/93198017327e240584facc073ae5853a1cc78a6c</a>

As stated in the spec [1], `template` tag start token should set framesetOk
flag to false when in in-body/in-head insertion mode.

[1] <a href="https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inbody">https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inbody</a>

* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(HTMLTreeBuilder::processTemplateStartTag): As above
* LayoutTests/fast/parser/resources/frameset-inside-template.html: Add Test Case Resource
* LayoutTests/fast/parser/disable-frameset-ok-flag-inside-template.html: Add Test Case
* LayoutTests/fast/parser/disable-frameset-ok-flag-inside-template-expected.txt: Add Test Case Expectation

Canonical link: <a href="https://commits.webkit.org/263851@main">https://commits.webkit.org/263851@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/237d07ff13abdbc236d86b5052e0b89d51eca619

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5822 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5990 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6179 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7382 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6211 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5823 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6213 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5957 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/8119 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5928 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5978 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5276 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7437 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3455 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5252 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13216 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5322 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5331 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7550 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5775 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4743 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5216 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/5182 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1394 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9337 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5577 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on LayoutTests/fast/parser/disable-frameset-ok-flag-inside-template.html:
```diff
@@ -0,0 +1,15 @@
+var test = async_test("A frameset tag inside template should be simply ignored.");
+var iframe = document.createElement("iframe");
+iframe.onload = function() {
+  document.body.removeChild(iframe);
```
It seems like this test is missing an assertion??

Ahmad-S792 replied:
I just took copy from 'source.chromium.org' and it seems that it does produce output without 'assertion' as well.

rniwa replied:
But this test is missing the check that frameset is getting ignored. We need something like this:
```js
test.step(() => {
    assert_equals(iframe.contentDocument.querySelector('frameset'));
    assert_equals(iframe.contentDocument.querySelector('template').content.querySelector('frameset'));
})
```

Ahmad-S792 replied:
^ Let me test it locally. Doing multiple patch testing locally. :-)

Ahmad-S792 replied:
@rniwa - I just reverted this change and then tested locally and it fails without this patch and gives following error: `FAIL: Timed out waiting for notifyDone to be called`, so I think without 'assert_equals' in the test, it is still working as intended and shows that it needs to be fixed.

rniwa replied:
ah, ok.

Ahmad-S792 committed:
A `frameset` tag inside `template` should be simply ignored

https://bugs.webkit.org/show_bug.cgi?id=256455

Reviewed by Ryosuke Niwa.

This patch aligns WebKit with Web-Spec [1] and Blink / Chromium.

Merge: https://chromium.googlesource.com/chromium/src/+/93198017327e240584facc073ae5853a1cc78a6c

As stated in the spec [1], `template` tag start token should set framesetOk
flag to false when in in-body/in-head insertion mode.

[1] https://html.spec.whatwg.org/multipage/parsing.html#parsing-main-inbody

* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(HTMLTreeBuilder::processTemplateStartTag): As above
* LayoutTests/fast/parser/resources/frameset-inside-template.html: Add Test Case Resource
* LayoutTests/fast/parser/disable-frameset-ok-flag-inside-template.html: Add Test Case
* LayoutTests/fast/parser/disable-frameset-ok-flag-inside-template-expected.txt: Add Test Case Expectation

Canonical link: https://commits.webkit.org/263851@main
